include/envoy/network/connection.h

```diff
@@ -294,6 +335,11 @@ class Connection : public Event::DeferredDeletable, public FilterManager {
    *         occurred an empty string is returned.
    */
   virtual absl::string_view transportFailureReason() const PURE;
+  /**
+   * @return DynamicSocketOptions& the instance for performing runtime configurable socket
+   *         options on the socket file descriptor
+   */
+  virtual Network::DynamicSocketOptionsPtr& getDynamicSocketOptionsPtr() PURE;
```

Comment:
Why does this return a shared ptr? It seems like it should just return a raw reference, or better should connection inherit from DynamicSocketOptions?

Reply:
Cool question (some learning for me here @mattklein123 :), been thinking about this ... )! 

With the advent of smart pointers. Why should we ever show the raw pointer and expose it to the risk of its management?

I can understand that it might be a bit expensive to return the shared_ptr. Is that the primary reason for optimizing this?

I am returning a reference to the shared_ptr; would this be that expensive as well as returning a copy of it or a raw pointer ?

But then again, it is a trade-off right ?

With respect to the or clause in your comment. I am fine with making this connection inherit it. But fully dont understand the scope of Connection to suggest that it might not be overkill. 

I will wait on a final remark, before I change.

Thanks!

Reply:
From an interface perspective it doesn't make any sense (that I can think of) to return an object with a different lifetime from the connection itself, and it's actually the case that the connection should support all of these options, so from that perspective I would make it inherit. If you want to have a connection implementation which does not implement this interface, then return it as an `absl::optitional<std::reference<>>`, and most implementations will just return `*this`.

Reply:
@mattklein123 there are few doubts that I am running into with the interpretation of your comment. Please correct me if I am wrong in understanding this.

1. When you suggest that Connection class inherit from DynamicSocketOptions class, all the underlying virtual functions need to be defined in Connection class. So far we have 4 virtual functions. However in future there could be many more. Essentially there is a setter and getter for each option. I understand that there are about 15 options. Therefore we would need to implement about 30 API in the connection class. I feel this could lead to complexity in writing tests(may be I am wrong?) ... Did you intend for this.

2. Currently the DynamicSocketOptionsImpl instance is only created during the construction of Connection instance. Thus is it not currently scoped to the lifetime of the Connection ? Are you intending to restrict the creation of this instance otherwise by design ?

Please let me know your thoughts.

Thanks!

Reply:
@alyssawilk @yanavlasov @mattklein123 verdict ?

Reply:
I don't think returning a shared_ptr here makes sense. How do you plan on using this interface? Is this the interface that will be exposed to HTTP filters? Probably not? If not I think these methods can be implemented directly on the connection interface?